File: leetcode/dp/2355_Maximum_Number_of_Books_You_Can_Take/codes/dp.py
from typing import List

class Solution:
    def maximumBooks(self, books: List[int]) -> int:
        n = len(books)
        dp = [0 for _ in range(n)]
        stack = []
        res = 0

        for i in range(n):
            while stack and books[stack[-1]] >= books[i] - (i - stack[-1]):
                stack.pop()

            if not stack:
                length = min(i + 1, books[i])
                first = books[i] - length + 1
                dp[i] = (books[i] + first) * length // 2
            else:
                j = stack[-1]
                length = i - j
                first = books[i] - length + 1
                dp[i] = dp[j] + (books[i] + first) * length // 2

            stack.append(i)
            res = max(res, dp[i])

        return res

File: leetcode/dp/2355_Maximum_Number_of_Books_You_Can_Take/codes/dp.c
long long maximumBooks(int* books, int booksSize) {
    long long* dp = calloc(booksSize, sizeof(long long));
    int* stack = malloc(booksSize * sizeof(int));
    int top = -1;
    long long res = 0;

    for (int i = 0; i < booksSize; ++i) {
        while (top >= 0 && books[stack[top]] >= books[i] - (i - stack[top])) {
            --top;
        }

        if (top < 0) {
            int len = books[i] < i + 1 ? books[i] : i + 1;
            int first = books[i] - len + 1;
            dp[i] = ((long long)(books[i] + first) * len) / 2;
        } else {
            int j = stack[top];
            int len = i - j;
            int first = books[i] - len + 1;
            dp[i] = dp[j] + ((long long)(books[i] + first) * len) / 2;
        }

        stack[++top] = i;
        if (dp[i] > res) res = dp[i];
    }

    free(dp);
    free(stack);
    return res;
}
